### ISSUE TITLE
International translation issues.

### ISSUE THREAD

**Issue by BengtGustafsson** [hadrielk/string-interpolation#2]
Is it possible to make f-literals translatable:

a) using unmodified gettext.
b) using modified gettext
c) Via a new T"" type of literal, i.e. a built in translation feature of C++.
d) By some other means.

If no solution is presented quite a lot of the audience for this feature may be lost.

**Comment by hadrielk** [comment]
I think if the goal is to not lose that audience, then using a modified gettext is not really applicable. It's implemented inside other libraries (e.g., glibc), not the stdlib.

So I think if we want to capture that audience, we'd have to make it work with (1) using an unmodified gettext.

**Comment by hadrielk** [comment]
BTW, regarding just handling `gettext()`...

Assuming we have a variadic `_FmtToPercent()` operator defined for converting the first-argument from fmt-encoding to percent-encoding, as described briefly in the draft proposal, then to use a `gettext()` is relatively trivial in terms of macros:

```c
#define CALL_GTEXT(s, ...) gettext(s) __VA_OPT__(,) __VA_ARGS__
#define FMT(...) CALL_GTEXT( _FmtToPercent( _Extract( _Concat( __VA_ARGS__ ) ) ) )

// usage
printf(FMT("name={name:s}, val={i:d}\n"));

// becomes:
printf(gettext("name=%s, val=%d\n"), name, i);
```

Or something like that - I haven't tried it.

---

The hard one is the gnu-lib extension `ngettext()`, I think?

**Comment by BengtGustafsson** [comment]
I tried to decipher your macros above. Why does FMT take variadic parameters when it only has one string literal as the only possible parameter? Is this for concatenating literals inside the FMT macro call? I guess so.

This solves the runtime issues but is it ok that the gettext extraction process sees the expressions, presuming that it is redirected to pick up strings in the FMT() macro calls. No it isn't: The english text in the database in your example would be `"name={name:s}, val={i:d}\n"` while the gettext function call provides `"name=%s, val=%d\n"` to look for in the database, drawing a blank.

To make this work the strings have to be massaged in parallel to what _FmtToPercent does by either the gettext scanner program or some postprocessing step. This is of course doable but requires a new gettext scanner, at least to handle a mixture of _() and FMT macros, whereof only the latter's strings are to me massaged. This differentiation is required to be able to do any type of code migration.

Apart from this it seems like translation is Linux's own version of dll hell where .mo files are dumped in the same directory for totally different applications, causing confusion when the same string turns up in different domains. So maybe it is time to come up with a better system, preferably portable.

**Comment by BengtGustafsson** [comment]
Another possibility is that _FmtToPercent returns a hash of the original format string instead of the % formatted string. Then the external gettext or postprocessing function can use the same hash-function to recreate the hash as the untranslated string, do the {} to % transformation and expression removal and then use the result as the English "translation" of the hash.

This would improve the gettext lookup speed in general but probably reduce it for the "english" locale as there is no lookup today. It saves some memory if the program is running in another language as only the hashes vs. the entire original strings need to be in memory.

Probably someone has done such tricks with gettext today already.

This seems a bit brittle though as you'd need the .mo files for English to even run the program at all.

**Comment by hadrielk** [comment]
> Why does FMT take variadic parameters when it only has one string literal as the only possible parameter? Is this for concatenating literals inside the FMT macro call? I guess so.

Yes, it's just in case the usage has multiple.

**Comment by hadrielk** [comment]
> This solves the runtime issues but is it ok that the gettext extraction process sees the expressions, presuming that it is redirected to pick up strings in the FMT() macro calls. No it isn't: The english text in the database in your example would be "name={name:s}, val={i:d}\n" while the gettext function call provides "name=%s, val=%d\n" to look for in the database, drawing a blank.


No, `gettext()` would actually see `"name=%s, val=%d\n"`, as shown in the "becomes" section of that example. Because the `_FmtToPercent()` would have already converted the result of `_Extract()` to percent-based encoding syntax.

**Comment by hadrielk** [comment]
Ugh, I forgot about the tooling around `gettext()`, such as the `xgettext` program.

I think `xgettext` will need to understand this new f-string mechanism to some degree, enough so that it can know what the generated post-interpolation format-string will be when it scans the source code files. That's... unfortunate.

**Comment by hadrielk** [comment]
> Assuming we have a variadic `_FmtToPercent()` operator defined for converting the first-argument from fmt-encoding to percent-encoding

I should note that the above was in order to handle the use-case of invoking `printf()`, or C-code in general.

For purely C++ code, it appears the tooling such as `xgettext` already supports the format-string syntax in C++20. So for use-cases involving `std::vformat()` and such, we don't need to convert to percent-encoding as the final syntax.

**Comment by BengtGustafsson** [comment]
Thanks for digging up the xgettext name, I wrote gettext extractor, but that was not clear enough.

[How do I get the vertical quoting line to the left... clearly not by just copy-pasting it. Below I added > manually...]
> For purely C++ code, it appears the tooling such as xgettext already supports the format-string syntax in C++20. So for 
> use-cases involving std::vformat() and such, we don't need to convert to percent-encoding as the final syntax.

As I can understand it xgettext does not have to do anything with a format string, it is up to the translator to not mess up the {} with or without numbers in the translated string. With f-literals in a compiled language this is not so, the expressions can't be left in as the string replacement to the foreign language happens at runtime...

**Comment by hadrielk** [comment]
> [How do I get the vertical quoting line to the left... clearly not by just copy-pasting it. Below I added > manually...]

Yup, adding a `>` and space before the thing you want to quote is how to do it . (see what it did to your comment when you did it above)

> As I can understand it xgettext does not have to do anything with a format string, it is up to the translator to not mess up the {} with or without numbers in the translated string. With f-literals in a compiled language this is not so, the expressions can't be left in as the string replacement to the foreign language happens at runtime...

Yes, exactly. `xgettext` needs to figure out the post-interpolated format-string - i.e. the one without expressions in it.

My comment was only about the fact that `xgettext` can handle C++20 brace-encoded format-strings, as opposed to only printf-style percent-encoded formatting strings. (originally I though it could only do the latter)
